### PR TITLE
Convert QuestionDefinition into an abstract class.

### PR DESCRIPTION
**Pull request by kendrickt**
### Description
Before, the QuestionDefinition had question type "text", and the other question types were sub-classed. After this change, every question type will be a subclass of the abstract class `QuestionDefinition` and a `TextQuestionDefinition` sub-class was added.

Jackson annotations are added to correctly serialize and deserialize `QuestionDefinition` into the proper sub-classes. 

Tests that used `QuestionDefinition` now use `TextQuestionDefinition`.

An `UnsupportedQuestionDefinitionException` is added for `switch(QuestionType)` statements. 

### Approvals

Looking for ALL

### Checklist
- [ ] Created tests which fail without the change (if possible)
- [ ] Extended the README / documentation, if necessary

### CODE REVIEW

File: universal-application-tool-0.0.1/app/services/question/QuestionDefinitionBuilder.java [hunk 1]
```diff
@@ -56,18 +56,20 @@ public QuestionDefinitionBuilder setQuestionType(QuestionType questionType) {
...
         return new NameQuestionDefinition(
             id, version, name, path, description, questionText, questionHelpText);
-      case TEXT: // fallthrough intended.
-      default:
-        return new QuestionDefinition(
+      case TEXT:
+        return new TextQuestionDefinition(
             id, version, name, path, description, questionText, questionHelpText);
+      default:
+        // TODO: throw a real exception.
```
Comment (carolinedanzi) [inline universal-application-tool-0.0.1/app/services/question/QuestionDefinitionBuilder.java:71]: nit: seems like you implemented this TODO and it can be removed

File: universal-application-tool-0.0.1/app/services/question/QuestionDefinition.java [hunk 2]
```diff
@@ -5,12 +5,23 @@
...
+@JsonTypeInfo(
+    use = JsonTypeInfo.Id.NAME,
+    include = JsonTypeInfo.As.PROPERTY,
+    property = "serialization_type")
+@JsonSubTypes({
+  @JsonSubTypes.Type(value = AddressQuestionDefinition.class, name = "address"),
+  @JsonSubTypes.Type(value = NameQuestionDefinition.class, name = "name"),
+  @JsonSubTypes.Type(value = TextQuestionDefinition.class, name = "text")
+})
+public abstract class QuestionDefinition {
```
Comment (carolinedanzi) [inline universal-application-tool-0.0.1/app/services/question/QuestionDefinition.java:24]: suggestion (blocking): `getQuestionType` and `getScalars` below should be abstract methods that are implemented by the subclasses

**Review by carolinedanzi — APPROVED** [review]

File: universal-application-tool-0.0.1/app/services/question/TextQuestionDefinition.java [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+      @JsonProperty("questionText") ImmutableMap<Locale, String> questionText,
+      @JsonProperty("questionHelpText") Optional<ImmutableMap<Locale, String>> questionHelpText) {
+    super(id, version, name, path, description, questionText, questionHelpText);
+  }
+
+  @Override
+  public QuestionType getQuestionType() {
+    return QuestionType.TEXT;
+  }
+}
```
Comment (AWBaumann) [inline universal-application-tool-0.0.1/app/services/question/TextQuestionDefinition.java:27]: This should also implement the following method:

```
 /** Get a map of scalars stored by this question definition. */
  @JsonIgnore
  public ImmutableMap<String, ScalarType> getScalars() {
    return ImmutableMap.of("text", ScalarType.STRING);
  }
```

...but if I recall we were going to make QuestionDefinition abstract in another PR, so you can do it then.

**Review by AWBaumann — APPROVED** [review]

File: universal-application-tool-0.0.1/app/services/question/QuestionDefinitionBuilder.java (hunk 1 above)
Comment (bion) [reply universal-application-tool-0.0.1/app/services/question/QuestionDefinitionBuilder.java:71]: Conversation was resolved while the TODO is still there?

**Review by bion — APPROVED** [review]

File: universal-application-tool-0.0.1/app/services/question/QuestionDefinitionBuilder.java (hunk 1 above)
Comment (kendrickt) [reply universal-application-tool-0.0.1/app/services/question/QuestionDefinitionBuilder.java]: i might be having git troubles. The tests are failing on origin but look fine on my end.

Comment (kendrickt) [reply universal-application-tool-0.0.1/app/services/question/QuestionDefinitionBuilder.java]: i might be having git troubles. The tests are failing on origin but look fine on my end. when i try to push it says everything is up-to-date.
